tests/test_utils.py

```diff
@@ -193,6 +206,31 @@ def test_extract_model(self):
 
         assert model == model_unwrapped
 
+    @require_tpu
+    @require_huggingface_suite
+    def test_extract_model_recursive_fsdpv2(self):
+        # Specifically tests for FSDPv2 extraction
+        # reported in https://github.com/huggingface/transformers/pull/29780
+        xr.use_spmd()
+        from transformers import AutoModelForCausalLM
+
+        model = AutoModelForCausalLM.from_pretrained("gpt2")
+        orig_state_dict_keys = list(model.state_dict().keys())
+        num_devices = xr.global_runtime_device_count()
+        xs.set_global_mesh(xs.Mesh(np.array(range(num_devices)), (num_devices, 1), axis_names=("fsdp", "tensor")))
+
+        def nested_wrap(model):
```

Comment:
Could you add a comment on why this is needed?

Reply:
Done, @shub-kris can you verify my reasoning behind it is correct?

Reply:
Yes, it's correct and to the point.